source/dev_guide/testing.rst

```diff
@@ -0,0 +1,73 @@
+Testing E3SM Diagnostics
+==============================================
+
+Unit tests
+----------
+
+Run all unit tests by doing the following:
+
+    .. code::
+
+        pip install . # Install your changes
+        ./tests/test.sh # Run all unit tests
+
+Automated tests
+---------------
+
+We have a Continuous Integration / Continuous Delivery (CI/CD) workflow.
+The unit tests are run automatically as part of this.
+
+Complete run test
+-----------------
+
+``tests/complete_run_v2_3_0.py`` checks the images generated by all diagnostics to
+see if any differ from expected.
+This test is not run as part of the unit test suite, because it relies on a large
+quantity of data found on LCRC (Anvil/Chrysalis).
+
+    .. warning::
+        You have to run this test manually. It is not run as part of the CI/CD workflow.
+
+If you've been developing code on a different machine, you can get the code to a LCRC
+machine by doing the following. First, push your code to GitHub.
+Then log into a LCRC machine and do the following:
+
+    .. code::
+
+        git fetch <repo-name> <branch-name> # Fetch the branch you just pushed
+        git checkout -b run-lcrc-test <repo-name>/<branch-name>
+
+Now that you have your changes on LCRC, enter your development environment. Then:
+
+    .. code::
+
+        pip install . # Install your changes
+        cd tests
+        python complete_run_v2_3_0.py
+
+If this test passes, there is no need to continue. If it fails however, that means
+your code has changed what the output looks like.
+
+If you expected your code to do this, then you need to update the expected images.
+
+    .. code::
+
+
+        ls image_check_failures # These are the images that differ from expected.
+        # <date> below should be in the form `yyyy_mm_dd`.
```

Comment:
@tomvothecoder Would it be better to use the commit hash (i.e., the hash of the latest commit on a developer's pull request) than the date? That way we can see what commit produced a certain set of expected images. However, the `<date>` approach is much easier to sort and understand.

Or is it even worth keeping older versions? If we don't keep older versions we can keep the paths set to one expected images directory, no edits required. However, it may be useful to keep a record of what the images looked like on previous runs.

Reply:
If we only use the `<date>` format, would there be naming collisions with previous runs that overwrite the contents of the folder? 
For example, if you run the test suite on a previous commit and on a new commit, would the generated files from the previous commit be overwritten by the new commit? If it does get overwritten, I might suggest using a timestamp  (e.g., `yyyy_mm_dd_hh_mm`) and/or hash of the commit that the test suite was on. Timestamp would allow uniqueness so you can keep previous runs on the same branch, and the hash of the commit can make it clearer which commit the test suite was run on.

I think it would be worth keeping older versions of test suites to see if there is any regression between commits and their associated code changes. This works hand-in-hand with labeling the commit hash mentioned above.

Reply:
@tomvothecoder Well if you run the complete-run test, nothing is automatically transferred to the expected images, so there'd be no automatic over-writing.

Maybe `yyyy_mm_dd_hash` would be good? But hashes are long -- maybe we could just list the first 7 characters like on https://github.com/E3SM-Project/e3sm_diags/commits/master?

@chengzhuzhang If we do `yyyy_mm_dd_hash`, I think just keeping the path set to `_latest` would be difficult, because then people would have to manually keep track of both the date and commit associated with the latest directory.

Reply:
I think `yyyy_mm_dd_<first 7 characters of hash>` makes the most sense as a suffix. I actually see the need to edit the expected_file paths as a positive in that it makes it clear in the pull request that this code changes the expected output.

If we're good with that, I will edit the docs accordingly and merge this.

Reply:
We may think about the workflow more. Something in my mind
```
generate `complete_run_latest.py`, point to `complete_run_v2.3.py`
If `complete_run_latest.py` need to be updated (i.e., new feature introduced):
     generate  new `complete_run_v.x.x.py`, `complete_run_latest.py`, point to `complete_run_vx.x.py`
     Create result directory `actual`_timestamp
             if the changes are expected: 
                  point `expected`_latest to `actual`_timestamp

if `complete_run_v2.3.py` stays the same (i.e., bug fixes or small changes):
     Create result directory `actual`_timestamp
           fix any unexpected diff. 
```
Something like this: a simple system to track change of the complete_run.py script and the results directory. We can add hash to the file/directory names if that's preferred.

Reply:
`complete_run_v2_3_0.py` isn't really being changed all that much, except for those paths. (And I suppose for each new release, we'll have a new `tests/complete_run_vMAJOR_MINOR_PATCH.py` and `examples/tutorial/tutorial_vMAJOR_MINOR_PATCH_all_sets_E3SM_machines.py`.) 

Perhaps `/lcrc/group/e3sm/public_html/e3sm_diags_test_data/unit_test_complete_run/` could have the subdirectories `previous_versions` and `latest`. `latest` could have a README specifying hash and date that `expected_images_all_sets.txt` and `all_sets` are from. Whenever `latest` is updated, developers will move the old ones into `previous_versions`.

Reply:
This sounds good. Maybe it is also good to add something like: copy the complete run of a new release to a new directory under /lcrc/group/e3sm/public_html/e3sm_diags_test_data/viewers/ with a standard filename (`_vMAJOR_MINOR_PATCH` as suffix)?